index.bs

```diff
@@ -9,8 +9,184 @@ Editor: Kent Tamura, Google https://www.google.com/, tkent@google.com
+  }
+}
+customElements.define('question-box', QuestionBox);
+&lt;/script>
+
+&lt;style>
+question-box::part(checkbox) { color: red; }
+question-box::part(checkbox):state(checked) { color: green; }
+&lt;/style>
+
+&lt;question-box>Continue?&lt;/question-box>
+&lt;/body>
+</pre>
+</div>
+
+
+Exposing custom element states {#exposing}
+============================
+
+Each <a>autonomous custom element</a> has <dfn>custom states</dfn>, a [=/set=]
+of strings. It's an empty set initially.
+
+<pre class=idl>
+partial interface ElementInternals {
+  [SameObject, PutForwards=value] readonly attribute DOMTokenList states;
+};
+</pre>
+
+The {{states}} IDL attribute returns a [=non-attribute DOMTokenList=] object
+associated with context object's
+<a href="http://html.spec.whatwg.org/C/#internals-target">target element</a>
+and its [=custom states=].
+
+<div class="example" id="ex-non-boolean-state">
+[=Custom states=] can expose boolean states represented by
+existence/non-exsitence of string tokens. If an author wants to expose a state
+which can have three values, it can be converted to three boolean states.
+For example, a state called <code>readyState</code> with <code>"loading"</code>,
+<code>"interactive"</code>, and <code>"complete"</code> values can be mapped to
+three boolean states, <code>"readyState_loading"</code>,
+<code>"readyState_interactive"</code>, and <code>"readyState_complete"</code>.
+
+<pre class="lang-js">
+// Change the readyState from anything to "complete".
+this.readyState = "complete";
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/remove}}</l>("readyState_loading", "readyState_interactive");
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/add}}</l>("readyState_complete");
+</pre>
+</div>
+
+## Non-attribute DOMTokenList ## {#sec-non-attribute-domtokenlist}
+
+This section defines a variant of {{DOMTokenList}}, called
+<dfn>non-attribute DOMTokenList</dfn>. It has the following differences from
+{{DOMTokenList}}.
+
+- A [=non-attribute DOMTokenList=] has an associated element and an associated
+    <a>token set</a>, but no attribute local name.
+
+- A [=non-attribute DOMTokenList=] object's
+    <a href="https://dom.spec.whatwg.org/#concept-dtl-update">update steps</a> are:
+
+    1. Recompute the style of the associated element.
+      <!-- TODO: How to represent this by spec terms? -->
+
+- A [=non-attribute DOMTokenList=] object's
+    <a href="https://dom.spec.whatwg.org/#concept-dtl-serialize">serialize steps</a>
+    are to return the result of running the <a>ordered set serializer</a> for
+    <a>token set</a>.
+
+- A [=non-attribute DOMTokenList=] object has no <a>attribute change steps</a>
+    for its associated element.
+
+- Setting the {{DOMTokenList/value}} IDL attribute of a
+    [=non-attribute DOMTokenList=] object must set <a>token set</a> to the given
+    value, <a lt="ordered set parser">parsed</a>.
+
+
+Selecting a custom element with a speicfic state {#selecting}
+============================
```

Comment:
The following comment is polish:

I think the general model for specifying selectors is to have some kind of general, semantic, mostly-language-agnostic description in the https://drafts.csswg.org/selectors-4/ spec, and then a specific implementation strategy saying what it matches in https://html.spec.whatwg.org/multipage/semantics-other.html#pseudo-classes . So you may want to separate this into two sub-sections, one that would fit well into the Selectors spec, and one that would fit well into HTML.

The HTML one is probably just the one sentence you already have ("An element matches..."). The Selectors one is the rest, but maybe with an extra sentence or two similar to what is currently in the introduction.

Reply:
Done.  Please re-check.